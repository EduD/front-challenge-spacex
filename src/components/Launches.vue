<template>
  <div class="flex items-center bg-gray-200">
    <div class="flex-1"></div>
    <div class="flex-1 text-gray-700 bg-gray-200 px-4 py-2 m-2">
      <h3 class="py-10 text-3xl">Last Lauches ðŸš€</h3>
      <div v-if="loading > 0">Loading...</div>
      <div v-else-if="launches" class="result apollo">
        <div
          v-for="(mission, index) in launches"
          :key="index"
          class="w-full h-auto mx-auto border-solid border-gray-500 rounded-lg bg-white shadow-lg px-5 pt-3 pb-5 mb-5 text-gray-800"
          style="min-width: 500px;"
          v-on:click="$modal.show('launch-detail'), launchDetail = mission"
        >
          <div class="w-full">
            <p class="text-xl text-indigo-500 text-left h-3 mb-5">{{ mission.mission_name }}</p>
            <p>{{ mission.details.length > 200 ? mission.details.substring(0, 200) + '...' : mission.details }}</p>
            <p class="text-right">{{ formatDate(mission.launch_date_local) }}</p>
          </div>
        </div>
      </div>
      <div v-else>No results found</div>
    </div>
    <div class="flex-1"></div>
    <custom-modal
      height="auto"
      style="overflow-y: scroll;"
      @closed="$emit('closed')"
      name="launch-detail"
    >
      <div v-if="launchDetail.links">
        <img :src="launchDetail.links.flickr_images" />
      </div>
      <p class="text-2xl text-indigo-500 text-center h-3 mb-5 py-5">{{ launchDetail.mission_name }}</p>
      <p class="text-xl text-gray-800 text-center mb-5">{{ launchDetail.details }}</p>
      <div class="flex items-center mb-5">
        <div class="flex-1"></div>
        <div class="flex-2 text-center">
          <button
            v-on:click="openLink(launchDetail.links.article_link ? aunchDetail.links.article_link : launchDetail.links.video_link)"
            class="flex 1 bg-gray-800 hover:bg-gray-500 text-white py-2 px-3 uppercase font-semibold text-2xs rounded"
            type="button"
          >See more</button>
        </div>
        <div class="flex-1"></div>
      </div>
    </custom-modal>
  </div>
</template>

<script>
import gql from 'graphql-tag'
import moment from 'moment'

moment.locale('pt')

export default {
  name: 'Lauches',

  data () {
    return { launchDetail: {}, launches: [], loading: 0 }
  },
  apollo: {
    launches: {
      query: gql`
        query {
          launchesPast(limit: 10) {
            mission_name
            details
            launch_date_local
            links {
              article_link
              video_link
              flickr_images
            }
          }
        }
      `,
      update: data => data.launchesPast,
      loadingKey: 'loading'
    }
  },
  methods: {
    formatDate (date) {
      return moment(date).format('DD/MM/YYYY')
    },
    openLink (link) {
      window.open(link, '_blank')
    }
  }
}
</script>

<style scoped>
</style>
