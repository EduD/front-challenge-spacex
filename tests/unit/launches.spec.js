import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import Launches from '@/components/Launches.vue'
const localVue = createLocalVue()

test('displayed last launches correctly with query data', () => {
  const wrapper = shallowMount(Launches, { localVue })
  wrapper.setData({
    launches: [
      {
        mission_name: "Starlink-13 (v1.0)",
        launch_date_local: "2020-10-18T08:25:00-04:00",
        links: {
          article_link: null,
          video_link: "https://youtu.be/UM8CDDAmp98",
          flickr_images: [
            "https://live.staticflickr.com/65535/50500804918_eb1187e1b2_o.jpg",
            "https://live.staticflickr.com/65535/50501674637_f16f528728_o.jpg",
            "https://live.staticflickr.com/65535/50501515611_2a3753bed1_o.jpg",
            "https://live.staticflickr.com/65535/50501674632_0d5276b1b5_o.jpg"
          ]
        },
        details: "This mission will launch the thirteenth batch of operational Starlink satellites, which are expected to be version 1.0, from LC-39A, Kennedy Space Center. It is the fourteenth Starlink launch overall. The satellites will be delivered to low Earth orbit and will spend a few weeks maneuvering to their operational altitude of 550 km. The booster for this mission is expected to land on an ASDS."
      },
    ],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('renders correctly when loading last launches', () => {
  const wrapper = mount(Launches, {
    mocks: {
      $apollo: {
        queries: {
          launches: {
            loading: true,
          },
        },
      },
    },
  })

  expect(wrapper.element).toMatchSnapshot();
})
