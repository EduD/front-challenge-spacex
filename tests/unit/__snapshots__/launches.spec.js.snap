// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayed last launches correctly with query data 1`] = `
<div
  class="flex items-center bg-gray-200"
>
  <div
    class="flex-1"
  />
   
  <div
    class="flex-1 text-gray-700 bg-gray-200 px-4 py-2 m-2"
  >
    <h3
      class="py-10 text-3xl"
    >
      Last Lauches ðŸš€
    </h3>
     
    <div
      class="result apollo"
    />
  </div>
   
  <div
    class="flex-1"
  />
   
  <custom-modal-stub
    classes=""
    clicktoclose="true"
    height="auto"
    maxheight="9007199254740991"
    maxwidth="9007199254740991"
    minheight="0"
    minwidth="0"
    name="launch-detail"
    overlaytransition="vm-transition--overlay"
    shiftx="0.5"
    shifty="0.5"
    style="overflow-y: scroll;"
    transition="vm-transition--modal"
    width="600"
  >
    <!---->
     
    <p
      class="text-2xl text-indigo-500 text-center h-3 mb-5 py-5"
    >
      
    </p>
     
    <p
      class="text-xl text-gray-800 text-center mb-5"
    >
      
    </p>
     
    <div
      class="flex items-center mb-5"
    >
      <div
        class="flex-1"
      />
       
      <div
        class="flex-2 text-center"
      >
        <button
          class="flex 1 bg-gray-800 hover:bg-gray-500 text-white py-2 px-3 uppercase font-semibold text-2xs rounded"
          type="button"
        >
          See more
        </button>
      </div>
       
      <div
        class="flex-1"
      />
    </div>
  </custom-modal-stub>
</div>
`;

exports[`renders correctly when loading last launches 1`] = `
<div
  class="flex items-center bg-gray-200"
>
  <div
    class="flex-1"
  />
   
  <div
    class="flex-1 text-gray-700 bg-gray-200 px-4 py-2 m-2"
  >
    <h3
      class="py-10 text-3xl"
    >
      Last Lauches ðŸš€
    </h3>
     
    <div
      class="result apollo"
    />
  </div>
   
  <div
    class="flex-1"
  />
   
  <!---->
</div>
`;
